readme.md

A console calculation game written in python that includes power-ups, streaks, points and other cool things! <br/>
Any platform other than Windows is in beta.



Installation
- download the install.bat for windows or install.sh for linux and save where you want the program to be stored
- run install.bat / install.sh <br/>

Updating - WIP

- download the update.bat for windows or install.sh for linux and save in the directory where the program is stored
- run update.bat or install.sh


changelog.md
*****************************
CHANGELOG (added as of 1.8)
*****************************

2.8:
	NEW, corrupt streak powers, have an equal chance of being obtained as an epic streak power
		The two streak powers are:
			halves your timeout time
			reduces your streak by a set amount, random between 1 and your current streak - 1
	NEW, your current streak will be shown durng a game in normal or challenger difficulty.
	Bug fixes:
		issue with answer being 2 operations (div and multi) when either number = 1, so neither number can = 1 now.

2.7.3
	timing changes for ease of access

2.7.2:
	removed win-sound
	device support (in beta):
		additional devices supported, check requirements.txt

2.7.1:
	bug fixes:
		rare streak  power (timer) stacking
	balance changes:
		starting points same for all difficulty (500 x streak stacks)
		Same for gain (+ 75) and reduction (-50) x streak stacks
		common streak power(1): +10 points --> +50 points gained per correct answer
		common streak power(2): +5% points --> +15% points gained per correct answer
		rare streak power(2): +2 seconds --> +3 seconds on timeout
		epic streak power(2): +100 points --> +500 points gained per correct answer
		legendary streak power(1): Also start with double the amount of points
		legendary streak power(2): Also gain +30% points when answered correctly

2.7:
	bug fixes:
		cases of streak not resetting even when an answer is incorrect
		chance streak power not loading properly
	NEW double the amount of streak powers for each rarity (streak powers pt2):
		common = + 5% points
		rare = timer length + 2 seconds
		epic = + 100 points per question when answered correctly
		legendary = Answer only one question and gain the same as what you would of gained if you answered 5 questions
	other changes:
		now a lot more common to get a streak power but there is still a chance you wont get one

2.6.1:
	visual changes:
		updated start up instructions to be more clear.
		extended length of all leaderboards
	technical changes:
		scores and points displayed as reals so game is compatible with future updates

2.6:
	streak powers(pt1):
		chance to gain certain powers when you are on a streak above 1 and below 10
		different rarities of streak powers. The rarer it is, the more powerful it is
	bug fixes:
		critical issue with when choosing to view challenger leaderboard at the end
	Timeout changes:
		all +  1 second

2.5.1:
	reworded some text and fixed some spelling errors, for clarity
	Your streak will reset after playing a game at a streak of 10
	better input support, e.g. you can now capitilse certain inputs
	coluring for certain things changed and timing improvments

2.5:
	NEW streak feature; ths is where the longer you play games without getting an answer incorrect, the more points you gain and lose
	NEW more info about the game mechs for clarity
	bug fixes:
		Inputting an invalid input during viewing a challenger leaderboard would cause it to incorrectly displayed
		Some text being incorrectly displayed
		Some sounds never being played
	Other changes:
		can no longer change difficulty after pressing play again unlesss you were in practice mode which then causes your streak to reset
		combined two screens

2.4.1:
	NEW: view how you gained and lost your points during a game of 'normal' mode
2.4:
	more sound has been implemented as well as some being removed
	unfortunatly wav files are buggy for some reason and so can't use as much sound as i want
	slight improvements to practice mode being displayed
	changed some inputs about how they are displayed
	challenger starting difficulty is now set to 150, prev 50
	NEW seperate leaderboard for the top 5 scores for challenger difficulty
	bug fixes:
		challenger points resetting to default every time a new round is started

2.3.1:
	practice mode improvments:
		can now choose any difficulty
		looks smoother
		inputs when not prompted will be flushed
2.3:
	NEW practice mode:
		practice the game without a time limit or points
		continues forever until you end it

2.2.1:
	reworked the highscore leaderboards
	entry to the challenge now reqs 'c', prev 'challenge'

2.2:
	challenger update:
		NEW difficulty: challenger; same as hard but the timeout is 3 seconds aposed to 5, accessed by choosing hard
		Also, point value changes for challenger only:
			starting points: 50
			reduction: 30
			gain: 10
	visual changes:
		highscore leaderboard displayed better and now only displays challenger highscores (for now, changes will be made)
	balance changes:
		points (excluding challenger):
			gain: 25 --> 30
			reduction: 20 --> 25
			initial points: 100 --> 150

2.1:
	NEW: various sounds have been added
	NEW: displays the newest implemented feature at the top the first leaderboard
	balance changes:
		timings nerfed:
			easy: 7.5 --> 10
			medium: 5.25 --> 7.5
			hard: 3.75 --> 5
		number gen range changes:
			easy:
				num1: (1(min) --> 125(max)) --> (25(min) --> 100(max))
				num2: (-15(min) --> 1(max)) --> (1(min) --> 25(max))
			medium:
				num1: (max 100) --> (max 150)
				num2: (min -20) --> (min -25)
			hard:
				num1: (max 100) --> (max 150)
	ease of access changes:
		certain timings have been changed
		when timing out changes:
			Displaying 'incorrect' --> displaying 'Timed out'
	visual changes:
		spacing at welcome menu removed so whole screen can be viewed
	
2.0.1:
	bug fixes:
		spacing issue at the end when resetting the leaderboard
	ease of access:
		you can now capitalise the y / n when wanting to play again

2.0:
	Balance changes:
		Timout changes (in seconds):
			easy: 6 --> 7.5
			medium: 4.5 --> 5.25
			hard: 3 --> 3.75
		Inputing an invalid answer now removes less points; 20 --> 5
		easy diificulty values of nums can be negative now (min -10)
		modulus option removed
		the second number will now always be smaller than the first no matter the difficulty
		timeout only costs you 10 (from 20) points
	LEADERBOARD RESET; due to current high scores being unobtainable in current version
	bug fixes:

practice.py
import random
import os
import time  
from colorama import init, Fore, Style
import platform
init()

def flush_input():
    try:
        import msvcrt
        while msvcrt.kbhit():
            msvcrt.getch()
    except ImportError:
        import sys, termios    #for linux/unix
        termios.tcflush(sys.stdin, termios.TCIOFLUSH)

oS = platform.system()

if oS == 'Windows':
	clear = lambda: os.system('cls')
else:
	clear = lambda: os.system('clear')
	
def getSum(num1, num2):
	return(num1 + num2)

def getDifference(num1, num2):
	return(num1 - num2)
	
def getProduct(num1, num2):
	return(num1 * num2)

def getQuotient(num1, num2):
	return(num1 / num2)

answer = "0"

print(Fore.GREEN + "1 = Easy")
time.sleep(0.2)
print(Fore.YELLOW + Style.DIM + "2 = Medium")
time.sleep(0.2)
print(Fore.RED + Style.NORMAL + "3 = Hard",Fore.GREEN + "")
time.sleep(0.2)

#asks for a difficulty to be selected with checks in place
while True:
	try:
		flush_input()
		difficulty = int(input("Choose a difficultly level to practice in (can be changed later): "))
		if difficulty > 3 or difficulty < 1:
			difficulty = int("f")
		break
	except ValueError:
		print(Fore.RED + "")
		print("Invalid, try again")
		 

while answer != "end":
	operationFinal = random.randint(0, 3)
	if difficulty == 1:
		num1 = random.randint(25,100) 
		num2 = random.randint(1,25) 

	elif difficulty == 2:
		num1 = random.randint(25, 150) 
		num2 = random.randint(-25, 25)

	elif difficulty == 3:
		num1 = random.randint(-50, 150) 
		num2 = random.randint(-100, -50)
	
	if operationFinal == 0:
		corAns = "addition"
		answer = getSum(num1, num2)
	elif operationFinal == 1:
		corAns = "subtraction"
		answer = getDifference(num1, num2)
	elif operationFinal == 2:
		corAns = "multiplication"
		answer = getProduct(num1, num2)
	elif operationFinal == 3:
		corAns = "division"
		answer = getQuotient(num1, num2)
			
	answer = round(answer)


	clear()
	print(Fore.YELLOW + "Type 'end' to finish")
	print("Press enter to skip a question")
	print(Fore.MAGENTA + "")
	print("Type 'd' to change the difficulty")
	if difficulty == 1:
		print("You are playing on",Fore.GREEN + "'easy'",Fore.MAGENTA + "difficulty")
	elif difficulty == 2:
		print("You are playing on",Fore.YELLOW +  "'medium'",Fore.MAGENTA +  "difficulty")
	else:
		print("You are playing on",Fore.RED + "'hard'",Fore.MAGENTA + "difficulty")
	print(Fore.CYAN + "------------------------------------------------------------------") 
	print(Fore.YELLOW + "The numbers",num1, "and",  num2, "turn into the number", answer)
	print()
	print(Fore.CYAN  + "1 = addition")
	print("2 = subtraction")
	print("3 = multiplication")
	print("4 = division (to nearest integer)")

	flush_input()
	choice = input(Fore.MAGENTA + "What operation has happened here: ")

	if choice == "1":
		choice = int(choice)
		choice = choice - 1
		if choice == operationFinal:
				clear()
				print(Fore.GREEN + " Correct!")
				 
				print()
				time.sleep(1.25)

		else:
			clear()
			print(Fore.RED + " Incorrect")
			 
			print()
			print(Fore.GREEN + "Correct answer:", corAns)
			time.sleep(1.7)


	elif choice == "2":
		choice = int(choice)
		choice = choice - 1
		if choice == operationFinal:
			clear()
			print(Fore.GREEN + " Correct!")
			 
			print()
			time.sleep(1.25)

		else:
			clear()
			print(Fore.RED + " Incorrect")
			 
			print()
			print(Fore.GREEN + "Correct answer:", corAns)
			time.sleep(1.7)

		
	elif choice == "3":
		choice = int(choice)
		choice = choice - 1
		if choice == operationFinal:
			clear()
			print(Fore.GREEN +" Correct!")
			 
			time.sleep(1.25)

		else:
			clear()
			print(Fore.RED + " Incorrect")
			 
			print()
			print(Fore.GREEN + "Correct answer:", corAns)
			time.sleep(1.7)

		
	elif choice == "4":
		choice = int(choice)
		choice = choice - 1
		if choice == operationFinal:
			clear()
			print(Fore.GREEN + " Correct!")
			 
			time.sleep(1.25)
		else:
			clear()
			print(Fore.RED + " Incorrect")
			 
			print()
			print(Fore.GREEN + "Correct answer:", corAns)
			time.sleep(1.7)
	
	elif choice == "":
		print()

	elif choice == "end":
		break
		clear()

	elif choice == "d":
		clear()
		print(Fore.GREEN + "1 = Easy")
		time.sleep(0.2)
		print(Fore.YELLOW + Style.DIM + "2 = Medium")
		time.sleep(0.2)
		print(Fore.RED + Style.NORMAL + "3 = Hard",Fore.GREEN + "")
		time.sleep(0.2)

		#asks for a difficulty to be selected with checks in place
		while True:
			try:
				flush_input()
				  
				difficulty = int(input("Choose a difficultly level to practice in (can be changed later): ")) 
				if difficulty > 3 or difficulty < 1:
					difficulty = int("f")
				break
			except ValueError:
				print(Fore.RED + "")
				print("Invalid, try again")
				 
	
	else:
		clear()
		print(Fore.RED + " Invalid")
		 
		print()
		print(Fore.GREEN + "Correct answer:", corAns)
		time.sleep(1.7)

		
